source/ConditionsStore.h

```diff
@@ -0,0 +1,187 @@
+
+	// Class for DerivedProviders, the (lambda) functions that provide access
+	// to the derived conditions are registered in this class.
+	class DerivedProvider
+	{
+	friend ConditionsStore;
+	friend ConditionEntry;
+	public:
+		// Functions to set the lambda functions for accessing the conditions.
+		void SetGetFun(std::function<int64_t(const std::string&)> newGetFun);
+		void SetHasFun(std::function<bool(const std::string&)> newHasFun);
+		void SetSetFun(std::function<bool(const std::string&, int64_t)> newSetFun);
+		void SetEraseFun(std::function<bool(const std::string&)> newEraseFun);
+
+	public:
+		// This is intented as a private constructor, only to be called from within
+		// ConditionsStore. But we need to keep it public because of how the
+		// DerivedProviders are emplaced in the providers-map-variable.
+		DerivedProvider(const std::string &name, bool isPrefixProvider);
+		
+	private:
+		std::string name;
+		bool isPrefixProvider;
+	
+		// Lambda functions for accessing the derived conditions, with some sensible
+		// default implementations;
+		std::function<int64_t(const std::string&)> getFun = [](const std::string &name) { return 0; };
+		std::function<bool(const std::string&)> hasFun = [](const std::string &name) { return true; };
+		std::function<bool(const std::string&, int64_t)> setFun = [](const std::string &name, int64_t value) { return false; };
+		std::function<bool(const std::string&)> eraseFun = [](const std::string &name) { return false; };
+	};
+
+	
+	// Storage entry for a condition. Can act as a int64_t proxy when operator[] is used for access
+	// to conditions in the ConditionsStore.
+	class ConditionEntry
+	{
+	friend ConditionsStore;
+	public:
+		// int64_t proxy helper functions. Those functions allow access to the conditions
+		// using `operator[]` on ConditionsStore.
+		operator int64_t() const;
+		ConditionEntry &operator=(int64_t val);
+		ConditionEntry &operator++();
+		ConditionEntry &operator--();
+		ConditionEntry &operator+=(int64_t val);
+		ConditionEntry &operator-=(int64_t val);
+	
+	private:
+		int64_t value = 0;
+		DerivedProvider *provider = nullptr;
+		// The exact name of the prefixed derived condition that we want to
+		// access using this ConditionEntry.
+		std::string prefixedProviderKey;
+	};
+	
+
+	// Input_iterator helper class to iterate over primary conditions.
+	// This can be used when saving primary conditions to savegames and/or
+	// for displaying some data based on primary conditions.
+	class PrimariesIterator
+	{
+		using CondMapItType = std::map<std::string, ConditionEntry>::const_iterator;
+	
+	public:
+		PrimariesIterator(CondMapItType it, CondMapItType endIt);
+		
+		// Iterator traits
+		using iterator_category = std::input_iterator_tag;
+		using value_type = std::pair<const std::string, int64_t>;
+		using difference_type = void;
+		using pointer = const std::pair<const std::string, int64_t>*;
+		using reference = std::pair<const std::string, int64_t>;
```

Comment:
This code is indeed not correct as it is currently implemented. I expect that we can fix this by returning pointers to ConditionEntries instead of the pairs that are used now, but for now I added some TODO texts to the code here to at least signal the issues. I also started [a new PR](https://github.com/endless-sky/endless-sky/pull/7075) to remove the last condition-iterators, after which the PrimariesIterator can also be fully removed.

I am a bit affraid that tehhowch might care. He has some ideas on conditions of which I'm not fully aware yet, and he did remark in the past that he saw a need to be able to iterate over all possible conditions that exist. I'm not sure if this remark still applies after this PR and after #7075.